src/Engines/ElasticsearchEngine.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace Laravel\Scout\Engines;
+
+use Laravel\Scout\Builder;
+use Elasticsearch\Client as Elasticsearch;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Support\Collection as BaseCollection;
+
+class ElasticsearchEngine extends Engine
+{
+    /**
+     * @var string $index
+     */
+    protected $index;
+
+    /**
+     * Create a new engine instance.
+     *
+     * @param  \Elasticsearch\Client  $elasticsearch
+     * @return void
+     */
+    public function __construct(Elasticsearch $elasticsearch, $index)
+    {
+        $this->elasticsearch = $elasticsearch;
+
+        $this->index = $index;
+    }
+
+    /**
+     * Update the given model in the index.
+     *
+     * @param  Collection  $models
+     * @return void
+     */
+    public function update($models)
+    {
+        $models->each(function ($model) {
+            $this->elasticsearch->index([
+                'index' => $this->index,
+                'type' => $model->searchableAs(),
+                'id' => $model->getKey(),
+                'body' => $model->toSearchableArray(),
+            ]);
+        });
+    }
+
+    /**
+     * Remove the given model from the index.
+     *
+     * @param  Collection  $models
+     * @return void
+     */
+    public function delete($models)
+    {
+        $models->each(function ($model) {
+            $this->elasticsearch->delete([
+                'index' => $this->index,
+                'type' => $model->searchableAs(),
+                'id'  => $model->getKey(),
+            ]);
+        });
+    }
+
+    /**
+     * Perform the given search on the engine.
+     *
+     * @param  Builder  $query
+     * @return mixed
+     */
+    public function search(Builder $query)
+    {
+        return $this->performSearch($query, array_filter([
+            'filters' => $this->filters($query),
+            'size' => $query->limit ?: 10000,
+        ]));
+    }
+
+    /**
+     * Perform the given search on the engine.
+     *
+     * @param  Builder  $query
+     * @param  int  $perPage
+     * @param  int  $page
+     * @return mixed
+     */
+    public function paginate(Builder $query, $perPage, $page)
+    {
+        $from = (($page * $perPage) - $perPage);
+
+        $result = $this->performSearch($query, [
+            'filters' => $this->filters($query),
+            'size' => $perPage,
+            'from' => $from,
```

Comment:
This is the only place where you use the $from variable, no need for it to be a local variable.